Issue 1713: Suppress trailing ? in URL strings if no parameters

labkey-kevink opened the pull request:
#### Rationale
With the experimental no-question-mark-url flag enabled, it isn't safe to concatenate parameters to the end of a URL.

#### Related Pull Requests
- https://github.com/LabKey/platform/pull/1603
- https://github.com/LabKey/platform/pull/1997

#### Changes
- don't use string concatenation to add parameters to URLs

labkey-kevink committed:
Issue 1713: Suppress trailing ? in URL strings if no parameters
- don't use string concatenation to add parameters to URLs

labkey-adam commented on testresults/src/org/labkey/testresults/view/longTerm.jsp:
```diff
@@ -109,9 +109,11 @@
                     var failure = failuresByFailToRunRatio[i];
-                    failureTableHTML += '<tr><td><a href="<%=h(new ActionURL(TestResultsController.ShowFailures.class, c))%>viewType=yr&failedTest='+failure[0]+'">'+failure[0]+'</a></td><td>'+failure[1]+'</td></tr>';
+                    url.searchParams.set('failedTest', failure[0]);
+                    failureTableHTML += '<tr><td><a href="' + url.toString() + '">'+failure[0]+'</a></td><td>'+failure[1]+'</td></tr>';
```
HTML encode url and failures?

labkey-kevink replied:
Fixed

labkey-adam commented:
Should we introduce a new `JspBase.jsUrl()` method to simplify this pattern? I'm thinking:

```<%=jsUrl(new ActionURL(ShowRunAction.class, c))%>```

instead of

```new URL(<%=q(new ActionURL(ShowRunAction.class, c).getURIString())%>)```

This continues the theme of using ActionURL as much as possible, reducing the number of `getURIString()` references to check/annotate.

labkey-adam reviewed (changes requested):
Couple things to consider

labkey-tchad committed:
Add PR validation workflow (#97)

labkey-susanh committed:
Add build.gradle file in anticipation of removal of modules/build.gradle (#99)

labkey-adam committed:
Use ActionURLs instead of Strings (#100)

vagisha committed:
Issue 42128: Update LINCS pipeline to create L2 GCT files on the MacCoss lab's Cromwell server (#98)

- Create L2 GCT on Cromwell server instead of on PanoramaWeb
- Register a SkylineDocumentImportListener, and queue lincs pipeline job in the onDocumentImport() method.
- API key is generated for the user before submitting Cromwell job and is deleted when the job completes.

labkey-tchad committed:
Update cromwell job to build in 21.2 (#103)

vagisha committed:
PanoramaPublic: misc. updates (#104)

- Check for spectral libraries built from Prosit predictions
- Require that submitter and lab head user accounts have first and last names. This is needed for PX submissions.
- Include lab head, when available, in confirmation email sent on data copy.
- Allow admin to override PX submission type (e.g. incomplete PX submission to complete submission)
- Include change log in the HTML summary and exported PX XML
- Export PX XML should work even if data is not valid for a PX submission (e.g. missing .wiff.scan files)
- Display submitter name in the Targeted MS Experiment webpart

labkey-kevink committed:
- use JspBase.jsURL to create browser URL objects
- a few more ActionURL usages into browser URL objects
- fixed html encoding

labkey-kevink commented:
I added the `jsURL()` method and it works, but IntelliJ doesn't give good code completion for the local url variable when we do this.  We can add a local type annotation to the block, but it gets a little ugly:

```
/** @type URL */
var url = <%=jsURL(new ActionURL(...))%>;
// IntelliJ can code complete the properties on the url variable now
```

labkey-adam commented:
Looks okay to me. Most (all?) of these could be simplified by using `urlFor(actionClass)`. Even better... another variant of your new method: `jsUrl(actionClass)`. Optional... but super cool :)
